src/providers/ldap/ldap_id_subid.c

```diff
@@ -0,0 +1,264 @@
+    struct tevent_req *subreq = NULL;
+    const char **attrs;
+    int ret;
+
+    ret = build_attrs_from_map(state, state->ctx->opts->subid_map,
+                               SDAP_OPTS_SUBID_RANGE, NULL, &attrs, NULL);
+    if (ret != EOK) {
+        tevent_req_error(req, ENOMEM);
+        return;
+    }
+
+    subreq = sdap_get_generic_send(state,
+                                   state->ev,
+                                   state->ctx->opts,
+                                   sdap_id_op_handle(state->op),
+                                   state->sdom->subid_ranges_search_bases[0]->basedn,
+                                   state->sdom->subid_ranges_search_bases[0]->scope,
+                                   state->filter,
+                                   attrs,
+                                   state->ctx->opts->subid_map, SDAP_OPTS_SUBID_RANGE,
+                                   dp_opt_get_int(state->ctx->opts->basic,
+                                                  SDAP_SEARCH_TIMEOUT),
+                                   false);
+    talloc_free(attrs);
+    if (!subreq) {
+        tevent_req_error(req, ENOMEM);
+        return;
+    }
+    tevent_req_set_callback(subreq, subid_ranges_get_done, req);
+}
+
+
+static void subid_ranges_get_done(struct tevent_req *subreq)
+{
+    struct tevent_req *req = tevent_req_callback_data(subreq,
+                                                      struct tevent_req);
+    struct subid_ranges_get_state *state = tevent_req_data(req,
+                                                     struct subid_ranges_get_state);
+    int dp_error = DP_ERR_FATAL;
+    int ret;
+    struct sysdb_attrs **results;
+    size_t num_results;
+
+    ret = sdap_get_generic_recv(subreq, state, &num_results, &results);
+    talloc_zfree(subreq);
+    if (ret) {
+        tevent_req_error(req, ret);
+        return;
+    }
+
+    ret = sdap_id_op_done(state->op, ret, &dp_error);
+    if (dp_error == DP_ERR_OK && ret != EOK) {
+        /* retry */
+        ret = subid_ranges_get_retry(req);
+        if (ret != EOK) {
+            tevent_req_error(req, ret);
+            return;
+        }
+        return;
+    }
+    state->sdap_ret = ret;
+
+    if (ret && ret != ENOENT) {
+        DEBUG(SSSDBG_OP_FAILURE, "Failed to retrieve subid ranges.\n");
+        state->dp_error = dp_error;
+        tevent_req_error(req, ret);
+        return;
+    }
+
+    if (num_results == 0 || !results) {
+        DEBUG(SSSDBG_MINOR_FAILURE,
+              "No such user '%s' or user doesn't have subid range\n",
+              state->name);
+        /* TODO: delete cached ranges? */
+    } else {
+        if (num_results > 1) {
+            DEBUG(SSSDBG_OP_FAILURE,
+                  "Multiple subid ranges, only first will be processed\n");
+        }
+
+        /* store range */
+        sysdb_store_subid_range(state->domain, state->name,
+                                state->domain->user_timeout,
```

Comment:
Introduce its own entry_cache_subid_timeout options?

Reply:
User's subid range feels and looks like other NSS attributes. Does it really make sense to have different expiration time?

Reply:
subid range may be for user or group, right? We have expiration time for user and for group so it may be confusing that group subid expires with users.

Reply:
In general, `user` might have defined a set of `subuid` and/or `subgid` ranges(*) (sub**u**id - a range of subordinate UIDs, sub**g**is - a range of subordinate GIDs). Those ranges (both types) can be used by this user for mapping in parent namespace. So both types of subid ranges are kind of user's attribute.

`group` (as NSS object) can't have any subids defined.


(*) In current implementation FreeIPA supports only two options:
 - no subid ranges defined for user
 - user has both subuid and subgid defined (with the same fixed size 65k and the same interval start, but this implementation detail doesn't matter much for us)